docs/terraform-integration-tests.md

```diff
@@ -0,0 +1,62 @@
+DC/OS is tested in TeamCity by several groups of integration tests using dcos-terraform and AWS spot instances.
+
+# Connecting to a test cluster
+
+Once the TeamCity job has completed, the DC/OS cluster that was created will be deleted. However, if the job is still running, you can connect to the cluster.
+
+Once `terraform apply` has completed, the job outputs to the job log the generated SSH key, cluster address, and node IP addresses as well as a bash one liner to write the SSH key to disk and SSH into the master node.
```

Comment:
> Once `terraform apply` has completed

This is regardless of whether DC/OS is "healthy" or not, right? (That's what we need for easy debugging of those problems that break cluster bootstrap, often happening when making patches in the bowels of DC/OS)

Reply:
Yes, as long as Amazon actually returns an IP address for a node to us, we can output this. Of course, if they can't even give that, then we wouldn't be able to.

Reply:
> we can output this.

We *can* or we *do*? Right now I have a build where I know there are problems for some systemd units to start, and I need to understand these problems, so I look into the build log to find SSH credentials. And don't find them.

https://teamcity.mesosphere.io/viewLog.html?buildId=2083882

Mhm.

Reply:
> Once `terraform apply` has completed,

I think as my example shows this is too late. Not sure where `terraform apply` is supposed to *end* but it seems to look for some "health signal" from DC/OS.

Reply:
Hmmm... you'll get it once terraform apply *finishes* in this case. Since terraform is still running, it's not output. Right now, once terraform completes we'll immediately tear it down since terraform completed with failure (and not run tests).

So, a couple of things:

* A two hour timeout is much too high, we should lower that.
* If we really, really want to, we can try to scrape the terraform logs and output it while terraform is still running.
* We also are working on a way for you to tell TeamCity that you want it to wait to delete your cluster DCOS-56075, which would make waiting until after terraform has completed okay.

Reply:
Also, also the IPs *are* output in the TeamCity log:

```
[10:34:13][Step 6/12] module.dcos.module.dcos-install.module.dcos-private-agents-install.null_resource.private-agents[1]: Provisioning with 'remote-exec'...
[10:34:13][Step 6/12] module.dcos.module.dcos-install.module.dcos-private-agents-install.null_resource.private-agents[1] (remote-exec): Connecting to remote host via SSH...
[10:34:13][Step 6/12] module.dcos.module.dcos-install.module.dcos-private-agents-install.null_resource.private-agents[1] (remote-exec):   Host: 52.27.240.233
[10:34:13][Step 6/12] module.dcos.module.dcos-install.module.dcos-private-agents-install.null_resource.private-agents[1] (remote-exec):   User: centos
[10:34:13][Step 6/12] module.dcos.module.dcos-install.module.dcos-private-agents-install.null_resource.private-agents[1] (remote-exec):   Password: false
[10:34:13][Step 6/12] module.dcos.module.dcos-install.module.dcos-private-agents-install.null_resource.private-agents[1] (remote-exec):   Private key: false
[10:34:13][Step 6/12] module.dcos.module.dcos-install.module.dcos-private-agents-install.null_resource.private-agents[1] (remote-exec):   SSH Agent: true
[10:34:13][Step 6/12] module.dcos.module.dcos-install.module.dcos-private-agents-install.null_resource.private-agents[1] (remote-exec):   Checking Host Key: false
[10:34:13][Step 6/12] module.dcos.module.dcos-install.module.dcos-private-agents-install.null_resource.private-agents[0]: Provisioning with 'remote-exec'...
[10:34:13][Step 6/12] module.dcos.module.dcos-install.module.dcos-private-agents-install.null_resource.private-agents[0] (remote-exec): Connecting to remote host via SSH...
[10:34:13][Step 6/12] module.dcos.module.dcos-install.module.dcos-private-agents-install.null_resource.private-agents[0] (remote-exec):   Host: 54.214.78.108
[10:34:13][Step 6/12] module.dcos.module.dcos-install.module.dcos-private-agents-install.null_resource.private-agents[0] (remote-exec):   User: centos
[10:34:13][Step 6/12] module.dcos.module.dcos-install.module.dcos-private-agents-install.null_resource.private-agents[0] (remote-exec):   Password: false
[10:34:13][Step 6/12] module.dcos.module.dcos-install.module.dcos-private-agents-install.null_resource.private-agents[0] (remote-exec):   Private key: false
[10:34:13][Step 6/12] module.dcos.module.dcos-install.module.dcos-private-agents-install.null_resource.private-agents[0] (remote-exec):   SSH Agent: true
[10:34:13][Step 6/12] module.dcos.module.dcos-install.module.dcos-private-agents-install.null_resource.private-agents[0] (remote-exec):   Checking Host Key: false
[10:34:13][Step 6/12] module.dcos.module.dcos-install.module.dcos-public-agents-install.null_resource.public-agents: Provisioning with 'remote-exec'...
[10:34:13][Step 6/12] module.dcos.module.dcos-install.module.dcos-public-agents-install.null_resource.public-agents (remote-exec): Connecting to remote host via SSH...
[10:34:13][Step 6/12] module.dcos.module.dcos-install.module.dcos-public-agents-install.null_resource.public-agents (remote-exec):   Host: 52.26.207.169
[10:34:13][Step 6/12] module.dcos.module.dcos-install.module.dcos-public-agents-install.null_resource.public-agents (remote-exec):   User: centos
[10:34:13][Step 6/12] module.dcos.module.dcos-install.module.dcos-public-agents-install.null_resource.public-agents (remote-exec):   Password: false
[10:34:13][Step 6/12] module.dcos.module.dcos-install.module.dcos-public-agents-install.null_resource.public-agents (remote-exec):   Private key: false
[10:34:13][Step 6/12] module.dcos.module.dcos-install.module.dcos-public-agents-install.null_resource.public-agents (remote-exec):   SSH Agent: true
[10:34:13][Step 6/12] module.dcos.module.dcos-install.module.dcos-public-agents-install.null_resource.public-agents (remote-exec):   Checking Host Key: false
[10:34:14][Step 6/12] module.dcos.module.dcos-install.module.dcos-private-agents-install.null_resource.private-agents[1] (remote-exec): Connected!
[10:34:14][Step 6/12] module.dcos.module.dcos-install.module.dcos-private-agents-install.null_resource.private-agents[0] (remote-exec): Connected!
[10:34:14][Step 6/12] module.dcos.module.dcos-install.module.dcos-public-agents-install.null_resource.public-agents (remote-exec): Connected!
[10:34:14][Step 6/12] module.dcos.module.dcos-install.module.dcos-private-agents-install.null_resource.private-agents[1]: Provisioning with 'remote-exec'...
[10:34:14][Step 6/12] module.dcos.module.dcos-install.module.dcos-public-agents-install.null_resource.public-agents: Provisioning with 'remote-exec'...
[10:34:14][Step 6/12] module.dcos.module.dcos-install.module.dcos-private-agents-install.null_resource.private-agents[1] (remote-exec): Connecting to remote host via SSH...
[10:34:14][Step 6/12] module.dcos.module.dcos-install.module.dcos-private-agents-install.null_resource.private-agents[1] (remote-exec):   Host: 52.27.240.233
[10:34:14][Step 6/12] module.dcos.module.dcos-install.module.dcos-private-agents-install.null_resource.private-agents[1] (remote-exec):   User: centos
[10:34:14][Step 6/12] module.dcos.module.dcos-install.module.dcos-private-agents-install.null_resource.private-agents[1] (remote-exec):   Password: false
[10:34:14][Step 6/12] module.dcos.module.dcos-install.module.dcos-private-agents-install.null_resource.private-agents[1] (remote-exec):   Private key: false
[10:34:14][Step 6/12] module.dcos.module.dcos-install.module.dcos-private-agents-install.null_resource.private-agents[1] (remote-exec):   SSH Agent: true
[10:34:14][Step 6/12] module.dcos.module.dcos-install.module.dcos-private-agents-install.null_resource.private-agents[1] (remote-exec):   Checking Host Key: false
[10:34:14][Step 6/12] module.dcos.module.dcos-install.module.dcos-public-agents-install.null_resource.public-agents (remote-exec): Connecting to remote host via SSH...
[10:34:14][Step 6/12] module.dcos.module.dcos-install.module.dcos-public-agents-install.null_resource.public-agents (remote-exec):   Host: 52.26.207.169
[10:34:14][Step 6/12] module.dcos.module.dcos-install.module.dcos-public-agents-install.null_resource.public-agents (remote-exec):   User: centos
[10:34:14][Step 6/12] module.dcos.module.dcos-install.module.dcos-public-agents-install.null_resource.public-agents (remote-exec):   Password: false
[10:34:14][Step 6/12] module.dcos.module.dcos-install.module.dcos-public-agents-install.null_resource.public-agents (remote-exec):   Private key: false
[10:34:14][Step 6/12] module.dcos.module.dcos-install.module.dcos-public-agents-install.null_resource.public-agents (remote-exec):   SSH Agent: true
[10:34:14][Step 6/12] module.dcos.module.dcos-install.module.dcos-public-agents-install.null_resource.public-agents (remote-exec):   Checking Host Key: false
[10:34:14][Step 6/12] module.dcos.module.dcos-install.module.dcos-private-agents-install.null_resource.private-agents[0]: Provisioning with 'remote-exec'...
[10:34:14][Step 6/12] module.dcos.module.dcos-install.module.dcos-private-agents-install.null_resource.private-agents[0] (remote-exec): Connecting to remote host via SSH...
[10:34:14][Step 6/12] module.dcos.module.dcos-install.module.dcos-private-agents-install.null_resource.private-agents[0] (remote-exec):   Host: 54.214.78.108
[10:34:14][Step 6/12] module.dcos.module.dcos-install.module.dcos-private-agents-install.null_resource.private-agents[0] (remote-exec):   User: centos
[10:34:14][Step 6/12] module.dcos.module.dcos-install.module.dcos-private-agents-install.null_resource.private-agents[0] (remote-exec):   Password: false
[10:34:14][Step 6/12] module.dcos.module.dcos-install.module.dcos-private-agents-install.null_resource.private-agents[0] (remote-exec):   Private key: false
[10:34:14][Step 6/12] module.dcos.module.dcos-install.module.dcos-private-agents-install.null_resource.private-agents[0] (remote-exec):   SSH Agent: true
[10:34:14][Step 6/12] module.dcos.module.dcos-install.module.dcos-private-agents-install.null_resource.private-agents[0] (remote-exec):   Checking Host Key: false
```

Reply:
From PM, @jgehrcke pointed out we still don't have the private key :)

Reply:
So I think what I'm going to do:

* Output the private key *before* terraform apply.
* Let the user collect the IPs from the terraform output and *document that that's where to look for them*
* Also output a nicely formatted step at the end.

Reply:
> Output the private key before terraform apply.

:heart:

Reply:
All set!

```
[12:09:11][Step 5/12] -----END RSA PRIVATE KEY-----
[12:09:11][Step 5/12] Process exited with code 0
[12:09:11]Step 6/12: terraform apply (Command Line) (running for 41s)
```

I'll move that into `terraform_init.sh` and update the docs in the morning.

Reply:
Thanks a bunch!